File: backend/services/config_builder.py
"""
Translate wizard selections into the engine's config dictionary.

The engine (data-pack-app/engine/generator.py) expects a specific config dict
structure. This module builds that dict from the frontend's GenerateRequest.
"""

from collections import OrderedDict
from openpyxl import load_workbook
from openpyxl.utils import column_index_from_string


def _detect_raw_data_frequency(filepath: str, sheet_name: str, date_col: str) -> str:
    """
    Auto-detect the raw data frequency by examining the date intervals.
    Returns 'monthly', 'quarterly', or 'annual'.
    """
    import datetime

    wb = load_workbook(filepath, read_only=True, data_only=True)
    ws = wb[sheet_name]
    col_idx = column_index_from_string(date_col)

    dates = []
    for row in ws.iter_rows(min_row=2, min_col=col_idx, max_col=col_idx,
                            values_only=True):
        val = row[0]
        if val is None:
            break
        if isinstance(val, (datetime.datetime, datetime.date)):
            dates.append(val if isinstance(val, datetime.date) else val.date())

    wb.close()

    if len(dates) < 2:
        return "annual"

    # Get sorted unique dates
    unique_dates = sorted(set(dates))
    if len(unique_dates) < 2:
        return "annual"

    # Average gap in days between consecutive unique dates
    gaps = [(unique_dates[i + 1] - unique_dates[i]).days
            for i in range(min(10, len(unique_dates) - 1))]
    avg_gap = sum(gaps) / len(gaps)

    if avg_gap < 45:
        return "monthly"
    elif avg_gap < 120:
        return "quarterly"
    else:
        return "annual"


def _get_unique_values(filepath: str, sheet_name: str, col_letter: str,
                       max_unique: int = 50) -> list[str]:
    """Read unique values from a column (for auto-generating filter breakouts)."""
    wb = load_workbook(filepath, read_only=True, data_only=True)
    ws = wb[sheet_name]
    col_idx = column_index_from_string(col_letter)

    values = set()
    for row in ws.iter_rows(min_row=2, min_col=col_idx, max_col=col_idx,
                            values_only=True):
        val = row[0]
        if val is None:
            break
        val_str = str(val).strip()
        if val_str:
            values.add(val_str)
        if len(values) >= max_unique:
            break

    wb.close()
    return sorted(values)


def build_engine_config(
    filepath: str,
    session_id: str,
    sheet_name: str,
    data_type: str,
    date_col: str,
    customer_id_col: str,
    arr_col: str,
    attributes: list[dict],          # [{"display_name": ..., "letter": ...}, ...]
    output_granularities: list[str],  # ["monthly", "quarterly", "annual"]
    fiscal_year_end_month: int,
    row_count: int,
    scale_factor: int,
) -> dict:
    """
    Build the config dict the engine expects.

    Parameters match the GenerateRequest schema fields.
    """
    # Build ordered attributes dict
    attrs = OrderedDict()
    for attr in attributes:
        attrs[attr["display_name"]] = attr["letter"]

    # Detect raw data frequency (this becomes time_granularity for the engine)
    raw_freq = _detect_raw_data_frequency(filepath, sheet_name, date_col)

    # Auto-generate filter breakouts from the first attribute
    filter_breakouts = []
    if attributes:
        first_attr = attributes[0]
        unique_vals = _get_unique_values(
            filepath, sheet_name, first_attr["letter"])
        for val in unique_vals:
            filter_breakouts.append({
                "title": val,
                "filters": {first_attr["display_name"]: val},
            })

    return {
        "raw_data_sheet": sheet_name,
        "raw_data_first_row": 2,
        "raw_data_last_row": row_count + 1,  # +1 because row 1 is the header
        "customer_id_col": customer_id_col,
        "date_col": date_col,
        "arr_col": arr_col,
        "attributes": attrs,
        "time_granularity": raw_freq,
        "output_granularities": output_granularities,
        "fiscal_year_end_month": fiscal_year_end_month,
        "scale_factor": scale_factor,
        "filter_breakouts": filter_breakouts,
        "data_type": data_type,
    }

File: frontend/src/engine/config_builder.ts
/**
 * Translate wizard selections into the engine's config dictionary.
 */
import type { Workbook } from 'exceljs';
import type { EngineConfig } from './types';
import { colNum } from './utils';

/**
 * Auto-detect the raw data frequency by examining date intervals.
 */
function detectRawDataFrequency(wb: Workbook, sheetName: string, dateCol: string, headerRow = 1): string {
  const ws = wb.getWorksheet(sheetName);
  if (!ws) return 'annual';

  const colIdx = colNum(dateCol);
  const dates: Date[] = [];

  ws.eachRow({ includeEmpty: false }, (row, rowNumber) => {
    if (rowNumber <= headerRow) return;
    const val = row.getCell(colIdx).value;
    if (val instanceof Date) {
      dates.push(val);
    }
  });

  if (dates.length < 2) return 'annual';

  const unique = [...new Set(dates.map(d => d.getTime()))].sort().map(t => new Date(t));
  if (unique.length < 2) return 'annual';

  const gaps: number[] = [];
  const limit = Math.min(10, unique.length - 1);
  for (let i = 0; i < limit; i++) {
    gaps.push((unique[i + 1].getTime() - unique[i].getTime()) / (1000 * 60 * 60 * 24));
  }
  const avgGap = gaps.reduce((a, b) => a + b, 0) / gaps.length;

  if (avgGap < 45) return 'monthly';
  if (avgGap < 120) return 'quarterly';
  return 'annual';
}

/**
 * Read unique values from a column (for auto-generating filter breakouts).
 */
function getUniqueValues(wb: Workbook, sheetName: string, colLetterStr: string, maxUnique = 50, headerRow = 1): string[] {
  const ws = wb.getWorksheet(sheetName);
  if (!ws) return [];

  const colIdx = colNum(colLetterStr);
  const values = new Set<string>();

  ws.eachRow({ includeEmpty: false }, (row, rowNumber) => {
    if (rowNumber <= headerRow) return;
    if (values.size >= maxUnique) return;
    const val = row.getCell(colIdx).value;
    if (val == null) return;
    const valStr = String(val).trim();
    if (valStr) values.add(valStr);
  });

  return [...values].sort();
}

export interface BuildConfigParams {
  wb: Workbook;
  sheetName: string;
  dataType: string;
  dateCol: string;
  customerIdCol: string;
  arrCol: string;
  attributes: { display_name: string; letter: string }[];
  outputGranularities: string[];
  fiscalYearEndMonth: number;
  rowCount: number;
  scaleFactor: number;
  dataFrequency?: string;
  // Cleaned table fields
  inputFormat?: 'raw' | 'cleaned';
  dateColumns?: string[];
  headerRow?: number;
  dateHeaderRow?: number;
}

/**
 * Build the config dict the engine expects.
 */
export function buildEngineConfig(params: BuildConfigParams): EngineConfig {
  const {
    wb, sheetName, dataType, dateCol, customerIdCol, arrCol,
    attributes, outputGranularities, fiscalYearEndMonth,
    rowCount, scaleFactor, dataFrequency,
    inputFormat, dateColumns, headerRow, dateHeaderRow,
  } = params;

  // Build ordered attributes dict
  const attrs: Record<string, string> = {};
  for (const attr of attributes) {
    attrs[attr.display_name] = attr.letter;
  }

  // Use user-provided frequency override if available, otherwise auto-detect
  const rawFreq = dataFrequency || detectRawDataFrequency(wb, sheetName, dateCol, headerRow);

  // Auto-generate filter breakouts from the first attribute
  const filterBreakouts: { title: string; filters: Record<string, string> }[] = [];
  if (attributes.length > 0) {
    const firstAttr = attributes[0];
    const uniqueVals = getUniqueValues(wb, sheetName, firstAttr.letter, 50, headerRow);
    for (const val of uniqueVals) {
      filterBreakouts.push({
        title: val,
        filters: { [firstAttr.display_name]: val },
      });
    }
  }

  return {
    raw_data_sheet: sheetName,
    raw_data_first_row: (headerRow || 1) + 1,  // data starts one row after headers
    raw_data_last_row: rowCount + (headerRow || 1),  // rowCount data rows after header
    customer_id_col: customerIdCol,
    date_col: dateCol,
    arr_col: arrCol,
    attributes: attrs,
    time_granularity: rawFreq,
    output_granularities: outputGranularities,
    fiscal_year_end_month: fiscalYearEndMonth,
    scale_factor: scaleFactor,
    filter_breakouts: filterBreakouts,
    data_type: dataType,
    input_format: inputFormat || 'raw',
    date_columns: dateColumns,
    date_header_row: dateHeaderRow,
  };
}
